Fix #9973: Change user settings model to support profiles

sajalasati opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #9973.
2. This PR does the following: 

- Introduces additional attribute `display_alias` needed to support profiles in the **UserSettingsModel** class.
- Modified test for `get_by_role` method in `UserSettingsModelTest` to ensure correctness of the output.
- Renamed other test functions of the class `UserSettingsModelTest` according to the rule no. 2 mentioned in [guidelines](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests)
- Changed the file `oppia_takeout_data.json` in `core/tests/data/oppia_takeout_data.zip` to include the additional attribute **display_alias**  introduced in `UserSettingsModel` in this PR.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

Commit:
added changes and tests

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
fix takeout service test

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

vojtechjelinek commented on core/storage/user/gae_models.py:
```diff
@@ -110,6 +110,22 @@ class UserSettingsModel(base_models.BaseModel):
     preferred_audio_language_code = ndb.StringProperty(
         default=None, choices=[
             language['id'] for language in constants.SUPPORTED_AUDIO_LANGUAGES])
+    # Displayed name for users on android. They are editable unlike usernames.
+    name = ndb.StringProperty(default=None)
+    # A code associated with profiles to provide a PIN based authentication
+    # within the account on android.
+    pin = ndb.StringProperty(default=None)
+    # List of profile user ids associated with a full user. None for profiles.
+    associated_profile_user_ids = ndb.StringProperty(repeated=True)
+    # Exploration language preferences specified by the user for android.
+    preferred_android_language_codes = ndb.StringProperty(
+        repeated=True,
+        indexed=True,
+        choices=[lc['code'] for lc in constants.SUPPORTED_CONTENT_LANGUAGES])
+    # Audio language preference used for audio translations on android.
+    preferred_android_audio_language_code = ndb.StringProperty(
+        default=None, choices=[
+            language['id'] for language in constants.SUPPORTED_AUDIO_LANGUAGES])
```
Are these only related to profiles? I think it would be useful to groupt the fields by the category of use: both profiles and accounts, only profiles, only accounts. Or this info should be always mentioned in the comment.

sajalasati replied:
Grouped them now.

vojtechjelinek commented on core/storage/user/gae_models_test.py:
```diff
@@ -95,24 +98,36 @@ def setUp(self):
             preferred_language_codes=self.GENERIC_LANGUAGE_CODES,
             preferred_site_language_code=(self.GENERIC_LANGUAGE_CODES[0]),
-            preferred_audio_language_code=(self.GENERIC_LANGUAGE_CODES[0])
+            preferred_audio_language_code=(self.GENERIC_LANGUAGE_CODES[0]),
+            name=(self.GENERIC_NAME),
+            pin=(self.GENERIC_PIN),
+            associated_profile_user_ids=(self.ASSOCIATED_PROFILE_IDS),
+            preferred_android_language_codes=(self.GENERIC_LANGUAGE_CODES),
```
```suggestion
            preferred_audio_language_code=self.GENERIC_LANGUAGE_CODES[0],
            name=self.GENERIC_NAME,
            pin=self.GENERIC_PIN,
            associated_profile_user_ids=self.ASSOCIATED_PROFILE_IDS,
            preferred_android_language_codes=self.GENERIC_LANGUAGE_CODES,
```

sajalasati replied:
Done.

vojtechjelinek reviewed (changes requested):
Thanks! Two comments.

BenHenning commented on core/storage/user/gae_models.py:
```diff
@@ -110,6 +110,22 @@ class UserSettingsModel(base_models.BaseModel):
     preferred_audio_language_code = ndb.StringProperty(
         default=None, choices=[
             language['id'] for language in constants.SUPPORTED_AUDIO_LANGUAGES])
+    # Displayed name for users on android. They are editable unlike usernames.
```
Is this specific to Android or might it be the case on web, as well? I wonder if we should call this 'profile_name'.

Do profiles also have usernames?

sajalasati replied:
Yes its specific to android only currently. On web we ask the user to set username and use it as a unique identifier. On android, all profiles can set their name (and edit them) and that is the displayed name there. We don't use username on android.
Yep, `profile_name` sounds better.

sajalasati replied:
Done. Used the `real_name` instead to indicate it will be there for full users as well as profiles.

BenHenning replied:
Sorry I'm a bit lost now--is this property going to be used specifically for profiles or for everyone? I'm also not keen on 'real name' as a name since we don't actually require that--people are free to use whatever name they'd like to here.

vinitamurthi replied:
I think we should call it profile_name. I think @sajalasati 's point is that even the Administrator profile on Android can rename this name and since administrator profile is a full user, this is being used for both full_users, and profiles.
But since that is actually a profile_name even for the administrator, we can set any android profile name related information in `profile_name`

sajalasati replied:
Done. Named the attribute to `profile_name` with comment explaining what it is.

BenHenning replied:
Makes sense--thanks for the context & rename.

BenHenning commented on core/storage/user/gae_models.py:
```diff
@@ -110,6 +110,22 @@ class UserSettingsModel(base_models.BaseModel):
             language['id'] for language in constants.SUPPORTED_AUDIO_LANGUAGES])
+    # Displayed name for users on android. They are editable unlike usernames.
+    name = ndb.StringProperty(default=None)
+    # A code associated with profiles to provide a PIN based authentication
```
Should this be part of ``UserAuthModel``, instead?

sajalasati replied:
I don't think so because it does not correspond to any authentication related information. It's just a user related information like user bio. And most importantly we can uniquely identify a person by their `name` because its meant to be unique only within the account, not across all users.

vojtechjelinek replied:
I too think that this should be part of `UserAuthModel`, also should the PIN be encrypted somehow?

sajalasati replied:
Sorry I am unable to understand this discussion. I was talking about `name` attribute, were both of you referring to `PIN` instead this whole time? 
If it's `PIN` then pls refer to this [comment](https://github.com/oppia/oppia/pull/10105#discussion_r463133394). If not, then please tell me more about the problem do you see.

BenHenning replied:
The comment was on line 115 which corresponds to the PIN. I'll follow up on that discussion thread.

sajalasati replied:
Done. Shifted `PIN` to `UserAuthModel` now in PR #10004.

BenHenning commented on core/storage/user/gae_models.py:
```diff
@@ -110,6 +110,22 @@ class UserSettingsModel(base_models.BaseModel):
+        repeated=True,
+        indexed=True,
+        choices=[lc['code'] for lc in constants.SUPPORTED_CONTENT_LANGUAGES])
+    # Audio language preference used for audio translations on android.
```
Why do we need new Android-specific language codes? Why not use the ones already defined for accounts?

sajalasati replied:
Once @vinitamurthi "suggested" me about this while designing the doc that keeping a separate field will allow users to have separate preferences for android and web.
WDYT about this @BenHenning?

vinitamurthi replied:
Hey @sajalasati , just to make it clear: That comment was only to show the trade-off between keeping them separate vs keeping them together. I think its important to understand the trade-off and then make a decision.

sajalasati replied:
I added a comment for them saying why I am including it. Trade-offs are just whether we want to provide separate preference choices on web and android or not. 
Although now it does feel more like a prod design concern to me. But apart from that, do you see any other issues that I might be face if I go with the current approach?

BenHenning replied:
My main concern is that it introduces complexity since we now have to keep two separate sets of properties in sync. Each time someone adds code to read/update this property, they need to think through why it's potentially different between web & Android (which it may not be). That adds maintenance complexity, and potentially a non-trivial amount.

vinitamurthi replied:
Based on the email discussion, I think we can share the audio language preference with web

sajalasati replied:
Done.

BenHenning commented on core/domain/takeout_service_test.py:
```diff
@@ -325,7 +328,13 @@ def set_up_non_trivial(self):
             preferred_language_codes=self.GENERIC_LANGUAGE_CODES,
             preferred_site_language_code=(self.GENERIC_LANGUAGE_CODES[0]),
-            preferred_audio_language_code=(self.GENERIC_LANGUAGE_CODES[0])
+            preferred_audio_language_code=(self.GENERIC_LANGUAGE_CODES[0]),
```
Nit: remove extra parentheses here, too.

sajalasati replied:
Done.

BenHenning reviewed (changes requested):
Thanks @sajalasati! Did a first pass. I have some high-level questions.

vinitamurthi commented on core/domain/takeout_service_test.py:
```diff
@@ -325,7 +328,13 @@ def set_up_non_trivial(self):
             preferred_site_language_code=(self.GENERIC_LANGUAGE_CODES[0]),
-            preferred_audio_language_code=(self.GENERIC_LANGUAGE_CODES[0])
+            preferred_audio_language_code=(self.GENERIC_LANGUAGE_CODES[0]),
+            name=(self.GENERIC_NAME),
+            pin=(self.GENERIC_PIN),
+            associated_profile_user_ids=(self.ASSOCIATED_PROFILE_IDS),
+            preferred_android_language_codes=(self.GENERIC_LANGUAGE_CODES),
```
I think you dont need any of these parenthesis

sajalasati replied:
Done.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -110,6 +110,22 @@ class UserSettingsModel(base_models.BaseModel):
+    # A code associated with profiles to provide a PIN based authentication
+    # within the account on android.
+    pin = ndb.StringProperty(default=None)
+    # List of profile user ids associated with a full user. None for profiles.
```
Write "None for profiles" as a fuller sentence -- "This is None if the user model corresponds to a ..."

sajalasati replied:
Done.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -110,6 +110,22 @@ class UserSettingsModel(base_models.BaseModel):
+    # List of profile user ids associated with a full user. None for profiles.
+    associated_profile_user_ids = ndb.StringProperty(repeated=True)
+    # Exploration language preferences specified by the user for android.
+    preferred_android_language_codes = ndb.StringProperty(
```
I thought I asked this before but why would these be different from preferred_language_codes, etc. above? The comments should explain.

sajalasati replied:
Explained in the comments now. Also pls refer to the discussion going on in above [comment thread](https://github.com/oppia/oppia/pull/10105#discussion_r462701426).

Commit:
addressed comments

Commit:
name changes

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
lint error fix

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -87,29 +131,13 @@ class UserSettingsModel(base_models.BaseModel):
-        default=None, choices=[
-            language['id'] for language in constants.SUPPORTED_AUDIO_LANGUAGES])
+    # List of profile user ids associated with a full user profile. This is
+    # None if the user model corresponds to a regular profile.
```
It seems to me a bit odd to describe "profiles" as "regular". To me, the regular case is the full user account. Also, what is an "irregular profile"?

In any case, can we pick a different descriptor and use that in general?

sajalasati replied:
Changed now. Now I have used the terms `Full user(primary profile)` and `secondary profile`. I hope that sounds much clearer.

vojtechjelinek commented on core/storage/user/gae_models.py:
```diff
@@ -44,36 +44,80 @@ class UserSettingsModel(base_models.BaseModel):
+    # A code associated with each user on android to provide a PIN based
+    # authentication to access a profile within the account. For a full user
+    # it provides access to administrator controls in the account (on android).
+    pin = ndb.StringProperty(default=None)
```
Should this be a part of `UserAuthModel`, also should this be somehow hashed or encrypted?

sajalasati replied:
Thanks for your comment!!
I think there's a confusion regarding the purpose of `UserAuthModel`. The `UserAuthModel` is only meant for storing information(metadata like `gae_id`) related to actual login process which involves calling an actual authentication service and doing the auth.
Whereas `PIN` is meant for in-account login - we can call it as profile login. So it's not connected to email with which the account was created or that sort of information. Like even when you are using the app in an offline mode, you can use the `PIN` based authentication - I am not sure how that is actually implemented in the android app, but it's totally an in-app thing and does not call any external service.

Regarding **hashing** it, that's a good suggestion. But do we need to consider about that in the storage itself or can I deal with it in later PRs - i.e do we want a class method for it implemented here or can I just implement it in domain layer - hash the PIN before updating it in the storage? My feeling is I can do it in domain layer.

BenHenning replied:
I was actually assuming that the PIN _is_ the authentication method for profiles, or rather, a combination of the parent user ID + PIN. By that thinking, a profile authenticates by providing a valid authenticated parent user ID & a PIN.

Re: hashing, I think that's a good idea. We don't let you recover PINs, only reset them, so we definitely don't need to store the actual PIN server-side. I don't think it's quite as sensitive as a password, but general hashing seems like a good idea.

sajalasati replied:
@BenHenning A combination of user ID + PIN will only be used actually. But I don't think we need to treat PIN as an authentication information to be sent along with the request to the server, as I am not planning to verify PIN at the backend.

The PIN (or its hash) is stored locally as well as on servers in the user data. Local PIN based auth system will ensure that only the correct user is making changes to that particular profile. Once that verification is done locally, PIN is no more used. At the backend we won't really check the PIN of the profile while processing its request, just use gae_id to verify that the request is coming from the correct account. We just use the gae_id, user_id to complete the request. Android side API can make sure to send the correct user id only of the profile -- For this particular reason I don't think we need to store PIN in `UserAuthModel`.

BenHenning replied:
How do we validate that a profile has access to certain APIs without verifying that it's logged in? I understand that we'll also be checking the PIN locally, but don't we also need to remotely validate the user in order for it to be considered 'logged in' for the purpose of retrieving data?

sajalasati replied:
@BenHenning Yes your right, we must do that verification too. Although it doesn't matter where we keep PIN for that auth purpose, but keeping it in the `UserAuthModel` only makes more sense since its an auth detail. 
I got a little lost into the details how the actual auth flow would look like and all, so didn't focus on the actual thing you were talking about which is just how we store it.
- Added `PIN` to the UserAuthModel in PR #10004 and removed it from `UserSettingsModel`.
- Hashing on `PIN` has been added as a point in the issue #10052 dealing with domain layer changes.

sajalasati replied:
Done.

BenHenning replied:
Thanks!

vojtechjelinek reviewed (changes requested):
Thanks! One more comment.

Commit:
removed unwanted attributes

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
fix merge conflict

sajalasati commented:
@nithusha21 PTAL! I have changed the file `oppia_takeout_data.json` in `core/tests/data/oppia_takeout_data.zip` to include the additional attributes `profile_name` and `associated_profile_user_ids` attributes added to `UserSettingsModel` in this PR.
This helped me resolve failing tests on:
- core.controllers.topic_viewer_test.TopicPageDataHandlerTests
- core.controllers.profile_test.EmailPreferencesTests
- core.controllers.profile_test.ExportAccountHandlerTests
- core.controllers.profile_test.FirstContributionDateTests

sajalasati commented:
@vojtechjelinek @BenHenning @seanlip @vinitamurthi PTAL! I have resolved the comments now.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -55,25 +57,52 @@ class UserSettingsModel(base_models.BaseModel):
+    subject_interests = ndb.StringProperty(repeated=True, indexed=True)
+    # When the user last edited an exploration.
+    # Exploration language preferences specified by the user.
+    # TODO(sll): Add another field for the language that the user wants the
```
Btw, I think you can drop this TODO. We seem to have it below already and we just never removed the TODO.

Keep the second sentence though ("These language preferences ...") -- just append it to the sentence before this one. It adds useful context.

sajalasati replied:
Done.

Commit:
documentation changes

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
revert time stamp in oppia_takeout_data.zip file

sajalasati commented:
@nithusha21 @vojtechjelinek Not sure which one of you to ask.

This is regarding the **[Failing backend test]** in this PR. I am a bit confused about it and not exactly sure what's happening.

In my **develop** branch on my local system (after doing a git fetch upstream; git reset --hard upstream/develop), the test `core.controllers.profile_test.ExportAccountHandlerTests.test_export_account_handler` is failing due to mismatch in values of `last_agreed_to_terms` and `last_logged_in` attributes in [this](https://github.com/oppia/oppia/blob/develop/core/controllers/profile_test.py#L890) particular assert condition. I haven't even touched the `core/controllers/profile_test.py` file. I have created a [gist](https://gist.github.com/sajalasati/2f6b7f22ee4d6320ecf9855eb3ddd1c7) to show the exact error log. The exact mismatch happening is:

| Attribute   |      Test Output      |  `core/tests/data/oppia_taekout_data.zip` |
|----------|:-------------:|------:|
| `last_agreed_to_terms` | 1558290600000.0 | 1558310400000.0 |
|last_logged_in |    1558290600000.0   | 1558310400000.0   |

In the `ExportAccountHandlerTests` class, we have [defined](https://github.com/oppia/oppia/blob/develop/core/controllers/profile_test.py#L804) `GENERIC_DATE = datetime.datetime(2019, 5, 20)` which translates to `1558290600000.0` in milliseconds - that is what the test function produces, but the value stored in `core/tests/data/oppia_taekout_data.zip` is something different. On using this timestamp only in the zip file, the tests pass locally. So I changed the timestamp, pushed my changes, but the test still failed with same error log as gist. It was surprising that the new attributes that I had added in the zip file were reflected, but not the timestamp change.

Can you please explain why `1558310400000.0` timestamp is being used? Also when the tests are failing for me locally, how were those changes even merged with the develop, wouldn't the tests have failed in other PRs too?

**Update**: The tests have now **passed** with the timestamp value same as original one i.e `1558310400000.0` in the `core/tests/data/oppia_taekout_data.zip` file.

vojtechjelinek commented:
@varun-tandon Hi, could you PTAL at Sajals last comment?

vojtechjelinek reviewed (approved):
LGTM!

varun-tandon commented:
@sajalasati @vojtechjelinek I think this is a time zone issue. If you compare the epoch that you got to the epoch that's in the zip file, you'll likely notice the time is the same if you adjust for time zones.

vojtechjelinek commented:
@varun-tandon How can this be mitigated, so that all devs get consistent behaviour of the tests?

varun-tandon commented:
@vojtechjelinek I'm working on another PR related to this, I'll include a fix in that PR.

sajalasati commented:
> @sajalasati @vojtechjelinek I think this is a time zone issue. If you compare the epoch that you got to the epoch that's in the zip file, you'll likely notice the time is the same if you adjust for time zones.

@varun-tandon Ah yes, I did see that the difference was less than a day. Thanks for clarifying!

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
Merge branch 'develop' into change-user-settings-model

nithusha21 reviewed (approved):
LGTM for the core/tests

BenHenning reviewed (approved):
Thanks @sajalasati --LGTM!

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -55,25 +57,51 @@ class UserSettingsModel(base_models.BaseModel):
+    last_logged_in = ndb.DateTimeProperty(default=None)
+    # Name of a user displayed in android UI. Unlike username, it can be
+    # edited and is unique only among the profiles of the account.
+    profile_name = ndb.StringProperty(default=None)
```
Let's call this display_alias for now. (There are other fields here which relate to the profile page and I want to try and avoid confusion with those.)

Rationale: I spoke to @BenHenning about this briefly and our thinking is that, in the long term, this field can also be used on web in the profile and author credits. But, want to try to find a name for the field that is not easily confused with `username` (hence the choice of a field name that doesn't have the word "name" in it).

sajalasati replied:
Made the changes now.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -87,29 +115,13 @@ class UserSettingsModel(base_models.BaseModel):
-        default=None, choices=[
-            language['id'] for language in constants.SUPPORTED_AUDIO_LANGUAGES])
+    # List of profile user ids associated with a full user (primary profile).
+    # This is None if the user model corresponds to a secondary profile.
```
Talked to @BenHenning -- let's drop this field. (He'll add more details to this.)

Also, I think the concept of "primary profile" and "secondary profile" is confusing. It also means that we have two names for the same concept: "full user" and "primary profile". How about "full user" and "profile user", from here on out? (Please update the doc and all other PRs/references accordingly -- @BenHenning agrees with these terms as well.)

BenHenning replied:
From talking to Sean, I realized that if we include the full user profile ID for profiles in UserAuthModel then we can retrieve all of the profiles associated with a given full user via a query (mitigating the need for this field entirely). The benefit of doing this is: we keep profiles more purely an actual Oppia account, and avoid potential confusion around full accounts having associated profile IDs (which is something we probably won't want in the long-term).

sajalasati replied:
@BenHenning I see just one drawback now - that this results in making an additional query in UserAuthModel now to fetch all profile ids when we need to operate on all of them. For eg. for a handler which Android app uses  to sync with the server i.e updates/fetches information for all the profiles on the server. This is was one of the things we considered in design doc also, but if we want to keep profiles more purely as an actual Oppia account, then I don't see any other way of doing it. This suggestion sounds good, but not sure if it would impact the performance a lot.

seanlip replied:
I think it should be fine. It's not going to be a very common query (like, on every page), is it?

sajalasati replied:
Corrected the naming of primary and secondary profile in code (this PR and PR #10004's UserAuthModel were only there). Will follow that for future PRs and the ones in progress too.

sajalasati replied:
@seanlip Yes it's not a very common query - basically happens only when we really need to fetch all the profiles. Talked to @BenHenning also today about it. It's fine.

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -94,25 +96,33 @@ def setUp(self):
             base_models.DELETION_POLICY.DELETE)
 
-    def test_apply_deletion_policy(self):
+    def test_apply_deletion_policy_registered_users_are_deleted(self):
```
Write in full sentences. E.g. test_that_apply_deletion_policy_deletes_registered_users().

sajalasati replied:
I think usage of full sentences in the naming of tests everywhere will make it very long (longer than the 80 characters limit). I have still updated the name and hope it looks better.

seanlip replied:
The name I suggested is only 1 character longer than the existing one, but if that causes issues, it is fine to omit "that".

If you write test names in small sentence fragments, it is hard to understand them. Past test names are not necessarily a good guide because we have only started being more rigorous about test names recently.

sajalasati replied:
Yes old tests are not good reference over what you are saying. Sorry I did not mean to ignore your suggestion. 
I was just confused which naming convention to follow - the one in the guide or what seems easy to read and understand. I myself preferred latter one. But I think the guide mostly wants to tell people not to miss any of the "3 things" in the name, their order should only be such that it makes it more readable.
I am pausing this change for a while, I have my 1:1 with @BenHenning today, I will get it clarified, and I will write the decision here and make the changes. @seanlip I hope you don't mind me doing that. I just wanted to understand it properly at once.

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -94,25 +96,33 @@ def setUp(self):
             user_models.UserSettingsModel.get_by_id(self.USER_1_ID))
-        # Test that calling apply_deletion_policy with no existing model
-        # doesn't fail.
+        user_models.UserSettingsModel.apply_deletion_policy(self.USER_2_ID)
```
Why add repeated tests for two more users, if it's not testing anything new? (That said, if each stanza does test something different, add a line comment above each stanza to explain what case it is testing, or alternatively separate the cases into separate tests with clear names.)

How about tests for deletion of profile users?

sajalasati replied:
For user2 -- it's marked as deleted=True. User3 is only different from user1 in the sense that it has more attributes None. Those attributes being not None could be linked to deletion of more storage models, hence kept separate tests. I will add a comment for each.

Thanks for pointing that out, I will create profile users and add tests for deleting them.

sajalasati replied:
Separated the tests for all kinds of users, commented properly as suggested.
(**Pending** adding tests for profile users)

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -94,25 +96,33 @@ def setUp(self):
+        self.assertIsNone(
+            user_models.UserSettingsModel.get_by_id(self.USER_3_ID))
+
+    def test_apply_deletion_policy_nonexistent_user_no_exception_raised(self):
```
test_no_exception_is_raised_when_nonexistent_user_is_deleted

sajalasati replied:
I was particularly following the convention given here - https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests Point (2). By following the order action-withCondition-outcome. And I have been trying to use that only everywhere. Which one should I proceed with? Same for [this](https://github.com/oppia/oppia/pull/10105/files#r464722524) comment.

seanlip replied:
Pick whichever one gives you something that people can understand. The overarching aim here is comprehensibility.

@BenHenning -- should we update the documentation?

sajalasati replied:
@seanlip Yes I agree, I just thought the order `action-withCondition-outcome` has to be followed.

sajalasati replied:
@seanlip I did not change it. It seemed readable to me. Also I looked at the other tests and they have used a similar approach, so I think it should be readable for anyone.

seanlip replied:
My original comment still stands.

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -121,23 +131,33 @@ def test_has_reference_to_user_id(self):
-        user = user_models.UserSettingsModel.get_by_role(
-            feconf.ROLE_ID_ADMIN)
-        self.assertEqual(user[0].role, feconf.ROLE_ID_ADMIN)
+    def test_get_by_role_for_admin_returns_user1_and_user3_only(self):
```
Talk about the behaviour of the test: ..._returns_only_admin_users

sajalasati replied:
Done.

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -121,23 +131,33 @@ def test_has_reference_to_user_id(self):
+            self.assertIn(user, actual_users)
 
-    def test_export_data_nonexistent_user(self):
+    def test_export_data_nonexistent_user_raises_exception(self):
```
Write complete sentences: test_export_data_for_nonexistent_...

sajalasati replied:
Done.

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -196,7 +220,7 @@ def test_get_new_id_method_returns_unique_ids(self):
             ids.add(new_id)
 
-    def test_create_raises_error_when_many_id_collisions_occur(self):
+    def test_get_new_id_forced_for_many_collisions_raises_error(self):
```
test_get_new_id_raises_error_if_too_many_collisions

sajalasati replied:
Done. Changed now.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -55,25 +57,51 @@ class UserSettingsModel(base_models.BaseModel):
+    # feconf.PROXIMAL_TIMEDELTA_SECS seconds.
+    last_logged_in = ndb.DateTimeProperty(default=None)
+    # Name of a user displayed in android UI. Unlike username, it can be
+    # edited and is unique only among the profiles of the account.
```
...among the profiles of the corresponding regular user account.

sajalasati replied:
Done.

BenHenning commented on core/storage/user/gae_models.py:
```diff
@@ -55,25 +57,51 @@ class UserSettingsModel(base_models.BaseModel):
+    # When the user last logged in. This may be out-of-date by up to
+    # feconf.PROXIMAL_TIMEDELTA_SECS seconds.
+    last_logged_in = ndb.DateTimeProperty(default=None)
+    # Name of a user displayed in android UI. Unlike username, it can be
+    # edited and is unique only among the profiles of the account.
```
```suggestion
    # Name of a user displayed in Android UI. Unlike username, it can be
    # edited and is unique only among the profiles of the account.
```

sajalasati replied:
Done.

BenHenning commented on core/storage/user/gae_models.py:
```diff
@@ -55,25 +57,51 @@ class UserSettingsModel(base_models.BaseModel):
+        default=None, choices=[
+            language['id'] for language in constants.SUPPORTED_AUDIO_LANGUAGES])
+
+
```
Nit: one line to separate attributes.

sajalasati replied:
Done.

Commit:
nit change and name correction

Commit:
test method name change

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

Commit:
fix linting

Commit:
changed method name

Commit:
removed profile ids list

Commit:
added test for profiles and corrected test names

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

sajalasati commented:
@BenHenning @seanlip PTAL now!
I have added all the required tests, renamed tests acc to the rule mentioned in guide (used complete sentences wherever possible), removed the `associated_profile_user_ids` attribute and did the required changes elsewhere. Also updated the PR description.
All of the changes demanded in the comments have been addressed. Please tell if it looks okay now. Thanks!

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -94,50 +105,77 @@ def setUp(self):
 
-    def test_apply_deletion_policy(self):
+    def test_apply_deletion_policy_for_registered_users_deletes_them(self):
+        # Case for a full user.
```
Where applicable, structure tests as follows: have a pre-check, then the test action, and then the post-check. The pre-check should be the same as the post-check.

So, here, for example, check first that the call results in "isNotNone", and then do the action, and then check "isNone". Ditto elsewhere. See point 1 [here](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests).

Please do this for tests in future PRs as well.

sajalasati replied:
Did it the way you described. Please check now.
Yes, noted, will do the same in future tests too.

seanlip commented on core/storage/user/gae_models_test.py:
```diff
@@ -196,7 +236,7 @@ def test_get_new_id_method_returns_unique_ids(self):
             ids.add(new_id)
 
-    def test_create_raises_error_when_many_id_collisions_occur(self):
+    def test_get_new_id_for_too_many_collisions_raises_error_(self):
```
Drop the underscore at the end.

sajalasati replied:
Sorry I missed that. Fixed now.

seanlip reviewed (changes requested):
Looks good! Just two minor comments -- thanks.

Commit:
address comments

Commit:
Merge remote-tracking branch 'upstream/develop' into change-user-settings-model

seanlip reviewed (approved):
Thanks @sajalasati! LGTM.

seanlip commented:
@jimbyo @vojtechjelinek could you please advise on the failing Lighthouse checks? If this is buggy then we need to revert the PR that caused it ASAP and send a note to oppia-dev.

/cc @sagangwee

jimbyo commented:
@seanlip @sajalasati Hey, sorry about this, the puppeteer script for updating the role to admin is being flaky right now. I'm making a fix to the flake in https://github.com/oppia/oppia/pull/10152/files. However, I could remove the flaky test now as a temporary fix ASAP. Rerunning the tests typically solves this error as well.

seanlip commented:
Hi @jimbyo -- thanks for the quick reply. Yes, please fix ASAP and file an issue for follow-up reinstatement of the test, so that we can unbreak develop. Also:

- In your fixup PR please mention which PR is being reverted.
- In the issue you file, please explain why the flake occurred and why it wasn't caught prior to merge.

Thanks!

BenHenning reviewed (approved):
Thanks @sajalasati! This LGTM.
